src/Device/Driver/AR62xx.cpp

```diff
@@ -0,0 +1,994 @@
+/*
+Copyright_License {
+
+  XCSoar Glide Computer - http://www.xcsoar.org/
+  Copyright (C) 2000-2016 The XCSoar Project
+  A detailed list of copyright holders can be found in the file "AUTHORS".
+
+  This program is free software; you can redistribute it and/or
+  modify it under the terms of the GNU General Public License
+  as published by the Free Software Foundation; either version 2
+  of the License, or (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program; if not, write to the Free Software
+  Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA  02111-1307, USA.
+}
+*/
+
+/*
+ * The driver is derived from the KRT2-driver. This version has implemented two methods yet:
+ * IMPLEMENTED > setting the active frequency
+ * IMPLEMENTED > setting the passive frequency
+ *
+ * TODO setting/reading dual scan on/off
+ * TODO reading frequency change on the radio
+ * TODO setting/reading squelsh settings
+ *
+ */
+
+#include "Device/Driver/AR62xx.hpp"
+#include "Device/Driver.hpp"
+#include "Device/Port/Port.hpp"
+#include "NMEA/Info.hpp"
+#include "RadioFrequency.hpp"
+#include "Thread/Cond.hxx"
+#include "Thread/Mutex.hxx"
+#include "Util/CharUtil.hxx"
+#include "Util/StaticFifoBuffer.hxx"
+#include "Util/Compiler.h"
+#include "LogFile.hpp"
+
+#include <cstdint>
+
+#include <stdio.h>
+
+//constants, expressions, defines for AR62xx radios binary protocol
+constexpr uint8_t HEADER_ID = 0xA5;
+#define PROTID 0x14;
+#define QUERY BIT(7)
+#define DUAL BIT(8)
+#define SQUELCH BIT(7)
+#define MAX_CMD_LEN 128
+#define ACTIVE_STATION 1
+#define PASSIVE_STATION 0
+#define BIT(n) (1 << (n))
+#define NAME_SIZE 30
+
+typedef union {
+  uint16_t intVal16;
+  uint8_t intVal8[2];
+} IntConvertStruct;
+
+typedef struct
+{
+  //active station frequency
+  double active_frequency;
```

Comment:
IMO floating point numbers is not a great choice for storing frequency, because they are not exact. E.g. `0.3` is actually `0.299999999999999988897769753748434595763683319091796875` when converted to decimal. You can get very weird results when manipulating these numbers. IMO fixed point decimal would fit much better here.

Reply:
That's why we have `class RadioFrequency` which encodes a radio frequency exact in a 16 bit integer.
https://github.com/XCSoar/XCSoar/blob/a8f6d1671fd69a73f4ed63c1de9301c0c5230cd8/src/RadioFrequency.hpp#L37-L45

Reply:
using RadioFrequency now :+1: